jcl/src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/abi/ProgrammableInvoker.java

```diff
@@ -0,0 +1,570 @@
+		return resultHandle;
+	}
+	
+	/* Obtain the filter that converts the passed-in argument to long against its type */
+	private static MethodHandle getArgumentFilter(Class<?> argTypeClass) {
+		/* Set the filter to null in the case of long by default as there is no conversion for long */
+		MethodHandle filterMH = null;
+		
+		if (argTypeClass == boolean.class) {
+			filterMH = booleanToLongArgFilter;
+		} else if (argTypeClass == char.class) {
+			filterMH = charToLongArgFilter;
+		} else if (argTypeClass == byte.class) {
+			filterMH = byteToLongArgFilter;
+		} else if (argTypeClass == short.class) {
+			filterMH = shortToLongArgFilter;
+		} else if (argTypeClass == int.class) {
+			filterMH = intToLongArgFilter;
+		} else if (argTypeClass == float.class) {
+			filterMH = floatToLongArgFilter;
+		} else if (argTypeClass == double.class) {
+			filterMH = doubleToLongArgFilter;
+		} else if (argTypeClass == MemoryAddress.class) {
+			filterMH = memAddrToLongArgFilter;
+		}
+		
+		return filterMH;
+	}
+	
+	/* The return value filter that converts the returned long value from the C function to the specified return type at Java level */
+	private MethodHandle getReturnValFilter(Class<?> returnType) {
+		MethodHandle filterMH = longObjToLongRetFilter;
+		
+		if (returnType == void.class) {
+			filterMH = longObjToVoidRetFilter;
+		} else if (returnType == boolean.class) {
+			filterMH = longObjToBooleanRetFilter;
+		} else if (returnType == char.class) {
+			filterMH = longObjToCharRetFilter;
+		} else if (returnType == byte.class) {
+			filterMH = longObjToByteRetFilter;
+		} else if (returnType == short.class) {
+			filterMH = longObjToShortRetFilter;
+		} else if (returnType == int.class) {
+			filterMH = longObjToIntRetFilter;
+		} else if (returnType == float.class) {
+			filterMH = longObjToFloatRetFilter;
+		} else if (returnType == double.class) {
+			filterMH = longObjToDoubleRetFilter;
+		} else if (returnType == MemoryAddress.class) {
+			filterMH = longObjToMemAddrRetFilter;
+		}
+		
+		return filterMH;
+	}
+	
+	/* The method (bound by the method handle to the native code) intends to invoke the C function via the inlined code */
+	Object runNativeMethod(long[] args) {
+		long returnVal = invokeNative(functionAddr.address().toRawLongValue(), cifNativeThunkAddr, args);
+		return Long.valueOf(returnVal);
+	}
+	
+	/* Verify whether the specified layout and the corresponding type are valid and match each other.
+	 * Note: will update after the struct layout (phase 2 & 3) is fully implemented.
+	 */
+	private void checkIfValidLayoutAndType(MethodType targetMethodType, FunctionDescriptor funcDesc) {
+		Class<?> retType = targetMethodType.returnType();
+		if (!validateArgRetTypeClass(retType) && (retType != void.class)) {
+			throw new IllegalArgumentException("The return type must be primitive/void or MemoryAddress" //$NON-NLS-1$
+												+ ": retType = " + retType);  //$NON-NLS-1$
```

Comment:
Excessive indentation (here and lines 406-407 and elsewhere).

Reply:
Fixed at all places as suggested above.